### PR TITLE
Make shortcodes work in `markdown` filter

### PR DESCRIPTION
**Pull request by hannobraun**
See #1350. Other details in the commits messages.

### CODE REVIEW

**Comment by southerntofu** [comment]
Hey thanks for the patch! That's was quick :)

I have two questions/critiques about it:

- do you think we can find a workaround to not reload all templates every time the MarkdownFilter is run (`filter()`)? that sounds very expensive, though i haven't done any benchmarking
- shortcodes usually [inherit](https://github.com/getzola/zola/blob/1ef8c85f53b4988fdafc0e6271cce590515d55aa/components/library/src/content/section.rs#L180) the page or section's rendering context, with your patch i believe shortcodes using such context will just crash (failed to access a missing variable), though i have not tested ; is that also your understanding?

**Comment by hannobraun** [comment]
Thanks for the review, @southerntofu! This is my first time diving into the innards of Zola, so your feedback is very helpful.

> do you think we can find a workaround to not reload all templates every time the MarkdownFilter is run (`filter()`)? that sounds very expensive, though i haven't done any benchmarking

Good point, I didn't think of that. I didn't notice any difference in my site, for what it's worth, but that isn't worth much :-)

> shortcodes usually [inherit](https://github.com/getzola/zola/blob/1ef8c85f53b4988fdafc0e6271cce590515d55aa/components/library/src/content/section.rs#L180) the page or section's rendering context, with your patch i believe shortcodes using such context will just crash (failed to access a missing variable), though i have not tested ; is that also your understanding?

I wasn't aware of that, but it certainly makes sense the way you explain it. I didn't encounter any errors building my site, but I don't think any of my shortcodes access such context.

I will look into both of these points, and will report back once I know more (probably within the next few days).

**Comment by hannobraun** [comment]
@southerntofu:
I've pushed two new commits. One extends the test site to demonstrate the error that you suspected would happen (and which is now confirmed). The other addresses your concern over reloading templates too often, which wasn't hard.

I'm still looking into approaches to address the error.

**Comment by hannobraun** [comment]
I've been thinking about how to solve the issue. The problem is that we need to somehow pass the current context into the markdown filter before it is run. I see multiple ways to do this, and none of them are any good:

1. Pass the context into `MarkdownFilter::new`. Right now, if I understand correctly, filters are [constructed on initialization](https://github.com/getzola/zola/blob/d734b1723c4097fe2bf5da1fe6b0f5c0060c6c2d/components/site/src/tpls.rs#L53-L56). We'd have to change that and construct a new `MarkdownFilter` every time we render a section or page. Sounds workable, but I'm not sure what the performance implications are, or if this would be undesirable for other reasons.
2. Pass the context into `tera::Filter::filter`. Since this isn't currently supported, we'd need to add an additional argument to the trait method, an associated type to the trait to specify the type of that argument, and some kind of mechanism in Tera to pass that argument to the filter. I don't think that this could be done in a backwards-compatible way (that would require associated type defaults, at least), and I'm not sure if such extensive changes to Tera would be welcome anyway.
3. Add an additional method (`filter_with_context`) to `tera::Filter` to pass dynamic context to. It might be possible to pass the context dynamically, as a `Box<Any>` and downcast it to the type that `MarkdownFilter` needs. I think this could be done in a backwards-compatible way (the new filter method could have a default implementation that just calls the old one), but would still require adding custom parameter infrastructure to Tera. The `Box<Any>` thing also feels like a hack.
4. Add a method to `MarkdownFilter` that accepts a context and call that whenever rendering a page. The problem here is that the filters live as `Arc<dyn Filter>` in `Tera`, so we'd at least need to add an `Any` bound to `Filter` to do anything with it. This would be an incompatible change to Tera and would limit which types could be used as filters (although I don't know if that would matter in practice).

Please let me know if I'm missing anything!

A much more practical solution would be to just merge this PR as-is, more of less. Yes, shortcodes in Markdown rendered by the `markdown` filter can't access page or section context, but given that without this PR the `markdown` filter doesn't render shortcodes at all, this is a strict improvement. Given that I've already spent more time on this than I wanted, and don't run into this limitation in my site, this is my preferred solution, to be honest.

**Comment by Keats** [comment]
> A much more practical solution would be to just merge this PR as-is, more of less.

I'm tempted to agree on that. Hopefully people using the `markdown` filter are not expecting a page or section argument since it can be used anywhere. As long as it's documented, I think it's fine.
The failing tests need to be fixed though

**Comment by hannobraun** [comment]
@Keats Great, I'll get the PR ready tomorrow, hopefully (next week, maybe).

The tests are failing because I demonstrated the shortcode context error in the test site, so I just have to drop that one commit to get them running again.

**Comment by southerntofu** [comment]
Hey, thanks for the in-depth analysis @hannobraun. I believe solution 2 would be viable given minor modifications. I wrote a proof-of-concept [here](https://github.com/southerntofu/zola/commit/929048375767388511e372ae3c3ccb95b6a7b072) which passes the test. i ended up passing from the template a list of (String, Value) (context elements)  but flattened as `Vec<Value>` directly to the markdown filter, using a dedicatd `context` argument. Some notes:

- tera currently has no way (that i could see) to designate the current context ; `__tera_context` is a pretty-printed string, so it's not suitable to be passed around
- tera has no Map literals and weird (broken?) List<List>> support
- it's just a proof of concept and i wouldn't recommend to adopt this list of context key/values to insert ; Map literals in tera (or another trick) would make it nicer
- i did not find a convenient way to write a unit test for the filter
- context has to be passed explicitly to the markdown filter ; i don't exactly like that (it's counter-intuitive from a template designer's perspective), but i appreciate the value of declaring what should be passed precisely and not cloning around a lot of data by default all the time

Any ideas?

Side note: there is a `json_encode` but i could not see a corresponding `json_decode` filter. The closest was `load_data` function but i don't think it supports a string source (that's not a path or a URL). I was thinking maybe there could be a way to turn `__tera_context` back into a Map after all, but well...

I'm happy with the PR as it is which is already a neat improvement @hannobraun, i'm merely continuing the conversation (which may be moved to a new issue).

About tera:

- there is already a ticket there about [support for Maps](https://github.com/Keats/tera/issues/420), however @Keats said it represented a considerable amount of work
- having tera filters aware of the context (your solution 3 or 4) is being discussed on [tera bugtracker](https://github.com/Keats/tera/issues/543#issuecomment-774698933) ; i believe it makes sense to have separate types of filters for functional filters (no context) and contextual filters

**Comment by Keats** [comment]
> there is already a ticket there about support for Maps, however @Keats said it represented a considerable amount of work

I'm holding off any work on Tera parser waiting for pest3 but it doesn't look like this is happening any time soon. I'll have to investigate other libraries but rewriting the parser is a lot of work.

> i believe it makes sense to have separate types of filters for functional filters (no context) and contextual filters

I think the context will be a function since there can be multiple contexts while rendering (eg a forloop might shadow some values from the context *inside* the loop). I don't think we want to distinguish between no context and with context though: the user can just ignore the context argument if they don't use it.

Tera v2 will basically be rewritten parser + context aware filters/functions. And if someone figures out how to avoid JSON serialisation it would probably be its own major version.

**Comment by southerntofu** [comment]
> waiting for pest3 but it doesn't look like this is happening any time soon

The maintainer [has said](https://github.com/pest-parser/pest/issues/489#issuecomment-775114684) they are too busy at the moment, but are happy to guide contributions ("I have a clear path forward if anyone would be interested to contribute"). So maybe passing around that call to contribution can help? We could have a new "Contributing" section in the README highlighting good-first-issues, issues we need help with, but also upstream issues and calls for maintainers. What do you think?

> I don't think we want to distinguish between no context and with context though

Well if you have clues how to make it zero-copy, sure. But as a template designer, i appreciate that how complex my build is depends on explicit trade-offs i make in the functions i call (eg. when i load information from other pages onto a page, i'm expecting a performance hit). If context passing is not zero-copy, i'm afraid trivial "functional templating" patterns (small macros everywhere) will be less attractive with tera.. but maybe that's a discussion for tera bugtracker?

About my poc context passing to the markdown filter, do you think:

- any other existing filters need to support that?
- if no Map literals are on the short-term TODO list, would a `newmap` function (or other map manipulation functions) be welcome either in zola or tera? ([this neater approach for example](https://github.com/southerntofu/zola/commit/9d5d93a02df7bf99e26bef04405657aab4a3126e) also passes the tests) this could include Map support in concat, or even an insert filter which works on both Lists and Maps (`{% set context = context | insert(key="foo", value="bar") %}`)
- current context could be exposed in a non-stringy way? (like go templates do with `.`) ; perhaps pretty printing structures could be a dedicated filter (or is that already what json_encode does?) so there would be no need to compute pretty-printed `__tera_context` on every render

**Comment by hannobraun** [comment]
I've update this pull request:
- Dropped the commit that demonstrated the shortcode context problem in the test site (tests pass again).
- Updated the respective commit message to mention the shortcode context limitation.
- Updated the commit that changes the documentation to mention the limitation.

Should be good to go now!

File: docs/content/documentation/templates/overview.md [hunk 1]
```diff
@@ -64,7 +64,8 @@ Zola adds a few filters in addition to [those](https://tera.netlify.com/docs/#fi
 in Tera.
 
 ### markdown
-Converts the given variable to HTML using Markdown. Shortcodes won't work within this filter.
+Converts the given variable to HTML using Markdown. Please note that shortcodes evaluated by this filter won't be able to access any context variables.
```
Comment (southerntofu) [inline docs/content/documentation/templates/overview.md]: I believe config context (and maybe some other things) will be available inside the markdown. Maybe something like:
```
Please note that shortcodes evaluated by this filter cannot currently access the current rendering context. `config` will be available, but accessing `section` or `page` (among others) from a shortcode called within the `markdown` filter will prevent your site from building. See [this discussion](https://github.com/getzola/zola/pull/1358).
```

Comment (hannobraun) [reply docs/content/documentation/templates/overview.md]: Done!

(Minus one word; "currently access the current" read a bit awkward, so I dropped "currently".)

**Comment by Keats** [comment]
> The maintainer has said they are too busy at the moment, but are happy to guide contributions ("I have a clear path forward if anyone would be interested to contribute"). So maybe passing around that call to contribution can help? We could have a new "Contributing" section in the README highlighting good-first-issues, issues we need help with, but also upstream issues and calls for maintainers. What do you think?

I've offered to help with the testing part since he said he had a plan for pest3 but I don't really know what that plan is and it isn't documented afaik. I don't know if the overlap Zola <-> Pest is big enough to be useful for the contribution bits. There's been discussion about pest3 since 2018 and at this point I would rather go with something more maintained I think.

> Well if you have clues how to make it zero-copy, sure. But as a template designer, i appreciate that how complex my build is depends on explicit trade-offs i make in the functions i call (eg. when i load information from other pages onto a page, i'm expecting a performance hit). If context passing is not zero-copy, i'm afraid trivial "functional templating" patterns (small macros everywhere) will be less attractive with tera.. but maybe that's a discussion for tera bugtracker?

It would have to be zero copy definitely. I have an idea but I don't know if it will actually work. If it does it would be trivial though but it would be a breaking change so for Tera v2.

I think your poc would be a bandaid and map literals/context should be fixed directly in Tera. Intermediate helpers like a new `insert` filter would probably cause too much churn imo.

**Comment by Keats** [comment]
Thanks @hannobraun !

**Pull request merged**